add <ShortcutPath> and NGEN info to VSIX packages

brettfo opened the pull request:
This PR is a continuation of #11661 except against `master` instead of `future`.  This also contains the name changes proposed over there.

FYI @dotnet/roslyn-ide, @dotnet/roslyn-infrastructure, @jmarolf, @TyOverby, @otawfik-ms, @shyamnamboodiripad, @Srivatsn, @basoundr, @Pilchie, @jasonmalinowski since they were explicitly tagged in the other PR.

brettfo commented:
test vsi please (just to make sure the packages still work on shipping versions of VS)

jasonmalinowski commented on src/VisualStudio/Setup/source.extension.vsixmanifest:
```diff
@@ -4,13 +4,38 @@
+      <Action Type="Ngen" Path="Microsoft.CodeAnalysis.VisualBasic.Features.dll" />
+      <Action Type="Ngen" Path="Microsoft.CodeAnalysis.VisualBasic.Workspaces.dll" />
+      <Action Type="Ngen" Path="Microsoft.CodeAnalysis.VisualBasic.EditorFeatures.dll" />
+      <Action Type="Ngen" Path="Microsoft.CodeAnalysis.EditorFeatures.Text.dll" />
```
Nit: can we sort this list so it's easier to find things?

brettfo replied:
Done.  Update coming shortly.

jasonmalinowski commented on src/ExpressionEvaluator/Package/source.extension.vsixmanifest:
```diff
@@ -4,6 +4,7 @@
     <Identity Id="21BAC26D-2935-4D0D-A282-AD647E2592B5" Version="|%CurrentProject%;GetBuildVersion|" Language="en-US" Publisher="Microsoft" />
     <DisplayName>Roslyn Expression Evaluators</DisplayName>
     <Description xml:space="preserve">Roslyn Expression Evaluators</Description>
+    <ShortcutPath>..\CommonExtensions\Microsoft\ManagedLanguages\VBCSharp\ExpressionEvaluators</ShortcutPath>
```
No ngen for the ExpressionEvaluator binaries?

brettfo replied:
Added.  Update coming shortly.

jasonmalinowski commented on src/VisualStudio/Setup/source.extension.vsixmanifest:
```diff
@@ -4,13 +4,38 @@
+      <Action Type="Ngen" Path="Microsoft.CodeAnalysis.VisualBasic.EditorFeatures.dll" />
+      <Action Type="Ngen" Path="Microsoft.CodeAnalysis.EditorFeatures.Text.dll" />
+      <Action Type="Ngen" Path="Microsoft.VisualStudio.LanguageServices.dll" />
+      <Action Type="Ngen" Path="Microsoft.CodeAnalysis.CSharp.dll" />
```
What about just the core CodeAnalysis.dll?

brettfo replied:
Good catch.  Added and update coming shortly.

jasonmalinowski commented on src/InteractiveWindow/VisualStudio/source.extension.vsixmanifest:
```diff
@@ -4,6 +4,7 @@
     <Identity Id="1F42C6D0-F876-4AF0-8185-1BEB0A325BB2" Version="|%CurrentProject%;GetBuildVersion|" Language="en-US" Publisher="Microsoft" />
     <DisplayName>VisualStudio Interactive Components</DisplayName>
     <Description>Interactive components for Visual Studio.</Description>
+    <ShortcutPath>..\CommonExtensions\Microsoft\ManagedLanguages\VisualStudioInteractiveComponents</ShortcutPath>
```
These are the bits used by non-ML too, right? Should this just be CommonExtensions\Microsoft\InteractiveWindow?

brettfo replied:
Good call.  Update pending.

jasonmalinowski commented:
I'm still terrified by the `..` that leads these, since that feels like we're trying to escape whatever folder we're supposed to be within. Is there a better way to express those?

brettfo commented:
@jasonmalinowski: updates pushed.

As for your question about `..`, that's the currently recommended way of doing this.  I'll reach out to the VSIX people and see if they can provide a `$(CommonExtensions)` variable (and others).

shyamnamboodiripad commented:
The `..` will probably look even worse once we have a compiler VSIX that tries to deploy into the MSBuild folder (which lives directly under VS Root install folder) :)

shyamnamboodiripad commented on src/InteractiveWindow/VisualStudio/source.extension.vsixmanifest:
```diff
@@ -4,6 +4,7 @@
     <Identity Id="1F42C6D0-F876-4AF0-8185-1BEB0A325BB2" Version="|%CurrentProject%;GetBuildVersion|" Language="en-US" Publisher="Microsoft" />
     <DisplayName>VisualStudio Interactive Components</DisplayName>
     <Description>Interactive components for Visual Studio.</Description>
+    <ShortcutPath>..\CommonExtensions\Microsoft\VisualStudioInteractiveComponents</ShortcutPath>
```
Just curious - Why aren't we ngen-ing the binaries in this VSIX? (Are they being ngen-ed in VS 2015?)

shyamnamboodiripad commented on src/VisualStudio/SetupInteractive/source.extension.vsixmanifest:
```diff
@@ -4,6 +4,7 @@
     <Identity Id="A5E5EE8D-39AE-42FF-8BBF-53C5B09C07D7" Version="|%CurrentProject%;GetBuildVersion|" Language="en-US" Publisher="Microsoft" />
     <DisplayName>Roslyn Interactive Language Services</DisplayName>
     <Description>Roslyn interactive language services.</Description>
+    <ShortcutPath>..\CommonExtensions\Microsoft\ManagedLanguages\VBCSharp\InteractiveLanguageServices</ShortcutPath>
```
Just curious - Why aren't we ngen-ing the Repl binaries in this VSIX? (Were they being ngen-ed in VS 2015?)

shyamnamboodiripad replied:
I think this VSIX (SetupInteractive) and the VSInteractiveComponents VSIX below should probably be merged into a single VSIX eventually.

@KevinH-MS I remember I had discussed this with you but can't seem to recall the conclusion :) Can these VSIXes be merged?

KevinH-MS replied:
Yes, they can be merged.

brettfo replied:
I'll add an item to my backlog to merge them, but for now @jmarolf and I decided it would be easier (e.g., fewer moving parts) to first get our existing VSIX packages working.

shyamnamboodiripad commented on src/ExpressionEvaluator/Package/source.extension.vsixmanifest:
```diff
@@ -4,13 +4,24 @@
     <Identity Id="21BAC26D-2935-4D0D-A282-AD647E2592B5" Version="|%CurrentProject%;GetBuildVersion|" Language="en-US" Publisher="Microsoft" />
     <DisplayName>Roslyn Expression Evaluators</DisplayName>
     <Description xml:space="preserve">Roslyn Expression Evaluators</Description>
```
This is fine for now. But will we need to remove the 'Roslyn' moniker from all these VSIX manifests at some point?

shyamnamboodiripad commented on src/VisualStudio/VisualStudioInteractiveComponents/source.extension.vsixmanifest:
```diff
@@ -4,13 +4,21 @@
+    <Actions>
+      <Action Type="Ngen" Path="Microsoft.CodeAnalysis.VisualBasic.InteractiveEditorFeatures.dll" />
+      <Action Type="Ngen" Path="Microsoft.CodeAnalysis.CSharp.InteractiveEditorFeatures.dll" />
+      <Action Type="Ngen" Path="Microsoft.CodeAnalysis.InteractiveEditorFeatures.dll" />
```
Should we also ngen the scripting binaries (MS.CA.Scripting, MS.CA.CSharp.Scripting)?

shyamnamboodiripad replied:
This VSIX also seems to be pulling in bunch of the same dlls as the Language Services VSIX (MS.CA, MS.CA.CSharp, workspaces, features etc.). This may be fine for now but I think we will need to avoid packaging the same binary in multiple VSIXes eventually.

shyamnamboodiripad replied:
BTW this VSIX includes MS.CA.VB.Scripting. Iirc, we should not be inserting this into VS, correct? cc @tmat

ManishJayaswal replied:
That's correct. We should not be inserting VB scripting.

shyamnamboodiripad replied:
That probably means we can't insert this VSIX as is. 
- We either need to remove the VB interactive stuff from the existing interactive VSIXes (which will be problematic for dogfooding) OR
- We need to create a new VSIX for the interactive components that excludes VB (and while we are here we can also merge the two VSIXes that we have for interactive today into one).

shyamnamboodiripad replied:
Based on offline conversation with @ManishJayaswal, it should be ok to remove the VB interactive / scripting binaries from this VSIX. This will probably cause some tests to fail - if so, we can disable the failing tests.

shyamnamboodiripad commented on src/VisualStudio/Setup/source.extension.vsixmanifest:
```diff
@@ -4,13 +4,33 @@
+      <Action Type="Ngen" Path="Microsoft.CodeAnalysis.VisualBasic.Workspaces.dll" />
+      <Action Type="Ngen" Path="Microsoft.CodeAnalysis.Workspaces.Desktop.dll" />
+      <Action Type="Ngen" Path="Microsoft.CodeAnalysis.Workspaces.dll" />
+      <Action Type="Ngen" Path="Microsoft.VisualStudio.LanguageServices.dll" />
```
Do we need to ngen MS.VS.LS.CSharp and MS.VS.LS.VB? Also MS.VS.LS.SlnExp and MS.VS.LS.Impl? I see these binaries when I unzip the VSIX...

Do we need to ngen MS.CA.Elfie (btw what that is for? :))

shyamnamboodiripad replied:
When I unzip them I also see that several of our VSIXes are pulling in some framework binaries (Sys.Collections.Immutable, Sys.Composition.*, Sys,Reflection.Metadata etc.). Is this ok?

Pilchie replied:
Oh.  Hmm, this might be a problem.  We currently get around that by having a list of binaries not to generate setup authoring for.  I believe we want them in the vsixes for dogfooding when we update them, but not in the vsix as part of a standard install.

@tmat probably knows more about this part.

tmat replied:
@Pilchie That is correct.

tmat replied:
Or maybe just a dogfooding-only vsix that contains all the dependencies.

brettfo commented:
To follow up with everybody, @jmarolf, @shyamnamboodiripad, and I have decided to continue with my changes in this PR and also remove VB interactive from the appropriate VSIX packages (I'll update this PR shortly).  This will enable us in the short term to get our setup working sooner.

For a longer-lead item we'll go back through and remove duplicate versions of System.Collections.Immutable, etc.

brettfo commented:
I've updated all assemblies included in the packages to be NGENed and removed VB scripting stuff.

And to find the likely test failures: test vsi please

brettfo commented:
Final call for comments/naming suggestions.

shyamnamboodiripad commented on src/VisualStudio/VisualStudioInteractiveComponents/source.extension.vsixmanifest:
```diff
@@ -4,13 +4,35 @@
+      <Action Type="Ngen" Path="System.Diagnostics.StackTrace.dll" />
+      <Action Type="Ngen" Path="System.IO.FileSystem.dll" />
+      <Action Type="Ngen" Path="System.IO.FileSystem.Primitives.dll" />
+      <Action Type="Ngen" Path="System.Reflection.Metadata.dll" />
```
I am not sure the System.* binaries like System.IO.* above need to be ngen-ed (since we would have been excluding them from the list at insertion-time before). It may be a good idea to check the exclusion list and only ngen binaries that we were actually dropping into PrivateAssemblies...

However, if this is harmless, I am also fine with leaving this in for now and coming back to clean this up later.

brettfo replied:
Good call.  I blindly added all assemblies that I found in the VSIX to this list.  I'll strip out the System.* ones.

tmat replied:
The facades are ngen'd in Dev15. See devenv.urt.config.tt

tmat replied:
Not sure how the system works, but shouldn't the responsibility for ngening CoreFX  assemblies be in a separate VSIX? I don't see why VisualStudioInteractiveComponents should be responsible for ngening them.  

```
    <Action Type="Ngen" Path="System.Collections.Immutable.dll" />
    <Action Type="Ngen" Path="System.Diagnostics.StackTrace.dll" />
    <Action Type="Ngen" Path="System.IO.FileSystem.dll" />
    <Action Type="Ngen" Path="System.IO.FileSystem.Primitives.dll" />
    <Action Type="Ngen" Path="System.Reflection.Metadata.dll" />
```

brettfo replied:
Regarding the CoreFX assemblies, yes, something else should be NGENing them; I'll remove them from this list.

As for the other CodeAnalysis dlls, ideally they wouldn't be included in this VSIX because they're in the other one (and will be NGENed there, too) but for this change we're going for the simplest delta possible to get us unblocked from our other work, but re-listing them here shouldn't cause a problem.  We'll be re-visiting our VSIX packages in the future to clean them up.

tmat replied:
Sounds good.

shyamnamboodiripad commented on src/VisualStudio/Setup/source.extension.vsixmanifest:
```diff
@@ -4,13 +4,46 @@
+      <Action Type="Ngen" Path="System.Composition.Hosting.dll" />
+      <Action Type="Ngen" Path="System.Composition.Runtime.dll" />
+      <Action Type="Ngen" Path="System.Composition.TypedParts.dll" />
+      <Action Type="Ngen" Path="System.Reflection.Metadata.dll" />
```
Same comment as below regarding ngening of System.* binaries...

shyamnamboodiripad commented on src/VisualStudio/Setup.Next/source.extension.vsixmanifest:
```diff
@@ -11,6 +11,11 @@
+      <Action Type="Ngen" Path="Microsoft.CodeAnalysis.EditorFeatures.Next.dll" />
+    </Actions>
+  </Installer>
   <Dependencies>
```
Do we also need to include ngen for Microsoft.VisualStudio.LanguageServices.Telemetry.vsix now that we are adding ngen for Setup.Next.vsix?

shyamnamboodiripad replied:
This VSIX (Setup.Next) doesn't have <ShortcutPath> specified. That means this will be extracted into a folder with mangled name, correct? Will ngen work in this case (i.e. don't we need well known path for ngen)?

brettfo replied:
I'll add NGEN for Telemetry.

Regarding Setup.Next, this package doesn't contain any assemblies, only pkgdef/redirects so there's nothing to NGEN so the mangled path is fine.

shyamnamboodiripad replied:
But looks like Setup.Next does contain Microsoft.CodeAnalysis.EditorFeatures.Next.dll. Does that mean that the `<Action Type="Ngen" Path="Microsoft.CodeAnalysis.EditorFeatures.Next.dll" />` line added above is not required?

brettfo replied:
That assembly is just a wrapper and doesn't need any NGEN.

shyamnamboodiripad commented:
Latest changes LGTM modulo last 4 comments that I added above.

brettfo commented:
PR updated again.  Any more takers, or can I merge this (once tests pass)?

<sub>test vsi please</sub>

brettfo commented:
test vsi please

brettfo commented:
test vsi please

brettfo commented:
test this please

brettfo commented:
Unit test failures are unrelated to this change and since I need to get a signed build of these VSIX packages to proceed, I'm merging them now.

Since VSI tests passed we should be good since those are the only things that would be affected by this change.
